### DISCUSSION TITLE
Document runtime representation of Gleam values

### DISCUSSION THREAD

**Discussion by giacomocavalieri** [gleam-lang/gleam#3306] [Ideas & suggestions]
Hello! Gleam's tour covers `@external` definitions by showing some JS usage examples creating a `DateTime`. While this might already cover the needs of most casual users I think it would be nice if we had an official documentation page somewhere covering more in-depth details about the runtime representation of things and how to write ffi modules.

Some ideas for the erlang target:
- the representation of types as tagged tuples
- show how gleam type names are turned into erlang atoms

Some ideas for the JS target:
- the representation of types as classes (with the exceptions of bools)
- show how `Nil` is turned into `undefined` (this is one of those things I always get stuck on)
- show how to import and use stdlib/prelude stuff from JS (it might be common wanting to return a result and it's not immediately clear how to do that)

I'm not really sure where this could fit (not really sure the tour is a great place for this kind of details) but it would be really valuable to have this officially documented somewhere. What do you think?

**Comment by Michael-Mark-Edu** [comment]
A potential idea I came up with is to have a link (to where?) that shows a cheat sheet of what everything in Gleam turns into for each target (maybe with some fancy color-coding to make it easy to distinguish the 3 languages at a glance). For example:
| Gleam | Erlang | JavaScript |
|--------|--------|--------|
| `Nil` | `nil` | `undefined` |
| `type MyType { MyConstructor(Int) }` | `-type my_type() :: {my_constructor, integer()}.` | `class MyConstructor extends $CustomType { constructor(x0) { super(); this[0] = x0; }}` |

Of course it would be formatted differently.

**Reply by giacomocavalieri** [reply]
Yeah I think something like this would be handy!

**Reply by lpil** [reply]
I would like a full guide rather than a cheat-sheet here

**Comment by lpil** [comment]
I would like to write a guide to FFI, though I have not got around to it myself.

**Comment by Michael-Mark-Edu** [comment]
I decided to start writing an FFI guide myself, and it's going well so far. Still early in development. I'm doing this mainly to learn TypeScript/React because that's what all the jobs want :)

I'll probably open the source code soon. I'm running it as a GitHub Pages site, and I'll likely use it for blogging in general.

**Reply by lpil** [reply]
Wonderful!!

**Reply by Michael-Mark-Edu** [reply]
Repo's public. You can check it out at https://michael-mark-edu.github.io/

It is far from complete though. Feedback would be welcome.

**Comment by Michael-Mark-Edu** [comment]
I found [this blog post](https://www.jonashietala.se/blog/2024/01/11/exploring_the_gleam_ffi/) which seems to do the same thing I was planning to do. My interest in this has waned, so I'd recommend directing people to that post for FFI information.
